Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp

```diff
@@ -26,16 +26,117 @@
+        };
+        m_callbacks.message = [](OpenCDMSession* session, const char message[]) {
+            OCDMSessionCallbacksNotifier::notify(session, &Observer::messageReceivedCallback, reinterpret_cast<const uint8_t*>(message), strlen(message));
+        };
+    }
+
+    static bool tryNotify(OpenCDMSession*, Notification, Ref<WebCore::SharedBuffer>&&);
+    static void notify(OpenCDMSession*, Notification, const uint8_t[], uint16_t);
+
+    friend class NeverDestroyed<OCDMSessionCallbacksNotifier>;
+    OpenCDMSessionCallbacks m_callbacks;
+    HashMap<OpenCDMSession*, Observer*> m_observers;
+    WebCore::GenericTaskQueue<WebCore::Timer> m_taskQueue;
+};
+
+OCDMSessionCallbacksNotifier& OCDMSessionCallbacksNotifier::singleton()
+{
+    static NeverDestroyed<OCDMSessionCallbacksNotifier> instance;
+    return instance;
+}
+
+void OCDMSessionCallbacksNotifier::addObserver(OpenCDMSession* session, Observer* observer)
+{
+    m_observers.set(session, observer);
+}
+
+void OCDMSessionCallbacksNotifier::removeObserver(Observer* observer)
+{
+    m_observers.removeIf([observer](const auto& keyAndValue) { return observer == keyAndValue.value; });
+}
+
+bool OCDMSessionCallbacksNotifier::tryNotify(OpenCDMSession* session, Notification method, Ref<WebCore::SharedBuffer>&& message)
+{
+    if (!isMainThread()) {
+        // Make sure all happens on the main thread to avoid locking.
+        callOnMainThread([session, method, message = WTFMove(message)]() mutable {
+            OCDMSessionCallbacksNotifier::tryNotify(session, method, WTFMove(message));
+        });
+        return true;
+    }
+
+    OCDMSessionCallbacksNotifier& self = OCDMSessionCallbacksNotifier::singleton();
+    auto entry = self.m_observers.find(session);
+    if (entry != self.m_observers.end()) {
+        (entry->value->*method)(String(message->data(), message->size()));
+        return true;
+    }
+
+    return false;
+}
+
+void OCDMSessionCallbacksNotifier::notify(OpenCDMSession* session, Notification method, const uint8_t message[], uint16_t messageLength)
+{
+    auto sharedMessage = WebCore::SharedBuffer::create(message, messageLength);
+    if (OCDMSessionCallbacksNotifier::tryNotify(session, method, WTFMove(sharedMessage)))
+        return;
+
+    // FIXME: There's no observer interested in the notification. It might be due to the fact it hasn't had chance
+    // to register yet because e.g. the notification is sent synchronously.
+    // Currently this happens for opencdm_create_session() which synchronously calls the process_challenge() callback and
+    // the only way to figure out who is the receiver is by OpenCMDSession. Problem however is that the caller can't know
+    // what's its sesssion at this point - it's creating one now.
+    auto& self = OCDMSessionCallbacksNotifier::singleton();
+    self.m_taskQueue.enqueueTask([session, method, sharedMessage = WTFMove(sharedMessage)]() mutable {
```

Comment:
What you commented about double moving here. Given that this happens inside the notifier class and it is controlled, I think we can keep it as a it is for efficiency reasons (we'd have to copy the shared buffer) if you add a comment about it. `WTFMove` is just a cast and the parameter is going to be effectively moved only when the first if returns `true`. When it return `false`, `tryNotify` hasn't touched the reference so even after being _casted for that_ it is not effectively moved. Tagging @zdobersek, my rvalue reference guru to double check my reasoning.

Anyway, I don't know if this piece of code will be here if you have to rework the notifier to avoid the fake parameters.

Reply:
I don't think this needs to be reworked and TBH what'd expect from move operation is to operator=(&&) to kick in effectively swapping the underneath storage.

Reply:
You're right. I checked with a simple program to simulate that

Reply:
Comment added